pkgs/development/libraries/valhalla/default.nix

```diff
@@ -0,0 +1,35 @@
+{ lib, stdenv, fetchFromGitHub, cmake, pkg-config
+, zlib, curl, protobuf, prime-server, boost, sqlite, libspatialite
+, luajit, geos, python3, zeromq }:
+
+stdenv.mkDerivation rec {
+  pname = "valhalla";
+  version = "3.1.0";
+
+  src = fetchFromGitHub {
+    owner = "valhalla";
+    repo = "valhalla";
+    rev = version;
+    sha256 = "04vxvzy6hnhdvb9lh1p5vqzzi2drv0g4l2gnbdp44glipbzgd4dr";
+    fetchSubmodules = true;
```

Comment:
We should de-vendor as much as possible like libosmium, protozero or rapidjson.

Reply:
I'd love to de-vendor things. Unfortunately upstream doesn't really support it in this case. Unlike some projects which include vendored sources for convenience but have the option to use system libraries instead, valhalla just assumes that the sources are there in third_party.

Taking rapidjson as an example, I can see lots of disadvantages to trying to de-vendor it:
* It's a header-only 'library', so there's no spacing-saving from using shared libraries, as there are no shared libraries.
* Future updates to the version of rapidjson in nixpkgs could break our build of valhalla (The valhalla devs will assume that rapidjson is the vendored version, so there won't be any conditional version-compatibility code to handle different versions, and there won't be a nice message saying, "Could not find a suitable version of rapidjson, we need version x.xx", it'll just assume everything's fine until it isn't).
* Valhalla 3.1.0 doesn't appear to build with the current version of rapidjson in nixpkgs
* Added complexity: we can't just put rapidjson in `buildInputs` because valhalla isn't looking for it. We have to manually symlink it into `third_party` in `postPatch` or similar.

The only minor advantage I can think of is:

* If a security vulnerability is discovered in rapidjson, then updating the nixpkgs rapidjson should cause a rebuild of valhalla that fixes it (or breaks valhalla).

libosmium is also a header-only library, so similar considerations apply.
protozero is... you guessed it... also a header only library!

I'm not always a fan of vendored dependencies, but I can understand why they do it. Vendoring gives upstream absolute control of the dependency version. This frees them from the obligation to handle different versions of the dependency, simplifying code (no `#if SOMELIB_VERSION >= 1234` littered throughout), and allowing greater confidence that their code is correct (smaller test space). Building a C++ package which vendors some of its header-only libraries is arguably no worse than packages using buildRustPackage, which vendor _all_ their rust dependencies. The fact that it's done via git submodules rather than cargo is just an implementation detail.

Reply:
> If a security vulnerability is discovered in rapidjson, then updating the nixpkgs rapidjson should cause a rebuild of valhalla that fixes it (or breaks valhalla).

That is the only part that really matters here and maybe nixpkgs specific changes to fix programs. We can overwrite versions if they break. The problem with vendoring is that many upstreams are lazy with updating and security updates take longer than necessary.